File: Angular/book.service.ts
import { Injectable } from '@angular/core';
import{ HttpClient } from '@angular/common/http'
import { Observable } from 'rxjs';
import { Book } from './book';
import { environment } from 'src/environments/environment';
@Injectable({
  providedIn: 'root'
})
export class BookService {

  private apiURL = environment.apiURL;

  constructor(private http:HttpClient) { }

  public getBooks():Observable<Book[]>{
    return this.http.get<Book[]>(`${this.apiURL}/api/books/getbooks`);
  }

  public regexTitle(title: string):Observable<any[]>{
    return this.http.get<any[]>(`${this.apiURL}/api/books/getbooks/${title}`);
  }

  public getGenre():Observable<any[]>{
    return this.http.get<any[]>(`${this.apiURL}/api/books/getbooks/getgenre`);
  }

  public getAuthor():Observable<any[]>{
    return this.http.get<any[]>(`${this.apiURL}/api/books/getbooks/getauthor`);
  }

  public getBookByGenre(genre: string):Observable<Book[]>{
    return this.http.get<Book[]>(`${this.apiURL}/api/books/getbooks/genre/${genre}`);
  }

  public getBookByAuthor(author: string):Observable<Book[]>{
    return this.http.get<Book[]>(`${this.apiURL}/api/books/getbooks/author/${author}`);
  }


  public getBookByTitle(title: string):Observable<Book[]>{
    return this.http.get<Book[]>(`${this.apiURL}/api/books/getbooks/title/${title}`);
  }

  public addBook(book: Book):Observable<Book>{
    return this.http.post<Book>(`${this.apiURL}/api/books/addbook`,book);
  }

  public deleteBook(bookID: number):Observable<void>{
    return this.http.delete<void>(`${this.apiURL}/api/books/deletebook/${bookID}`);
  }
}

File: Java/Book/BookService.java
package com.example.Bookstoredb.bookstore.books;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
public class BookService {

    private final BookRepository bookRepository;

    @Autowired
    public BookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    public List<Book> getBooks(){
        return bookRepository.findAllByOrderByBookIDAsc();
    }

    public Set<String> getGenre(){
        Set<String> genres = new HashSet<String>();
        List<Book> book = this.getBooks();
        for (Book value : book) {
           // assert genres != null;
            genres.add(value.getGenre());
        }
        return genres;
    }

    public Set<String> getAuthors(){
        Set<String> authors = new HashSet<String>();
        List<Book> book = this.getBooks();
        for (Book value : book) {
            authors.add(value.getAuthor());
        }
        return authors;
    }

    public List<Book> getBookByGenre(String genre){
        return bookRepository.findByGenre(genre);
    }
    public List<Book> getBookByAuthor(String author){
        return bookRepository.findByAuthor(author);
    }

    public List<Book> getBookByTitle(String title){
        return bookRepository.findByTitle(title);
    }

    public void addBook(Book book){
        bookRepository.save(book);
    }

    public void deleteBook(Long bookID){
        bookRepository.deleteById(bookID);
    }
}
